[Serializer] Improve ObjectNormalizer performance

dunglas opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #16179 |
| License | MIT |
| Doc PR | n/a |

Cache attributes detection in a similar way than in #16294 for PropertyAccess.

As for the PropertyAccess Component, I'll open another PR (in 2.8 or 3.1) allowing to cache these attributes between requests using Doctrine Cache.

@Tobion, can you try this PR with your benchmark to estimate the gain?

dunglas committed:
[Serializer] Improve ObjectNormalizer performances

iltar commented on src/Symfony/Component/Serializer/Normalizer/ObjectNormalizer.php:
```diff
@@ -159,4 +129,55 @@ public function denormalize($data, $class, $format = null, array $context = arra
+                } elseif (strpos($name, 'is') === 0) {
+                    // issers
+                    $attributes[lcfirst(substr($name, 2))] = true;
+                }
```
Aren't the constructor checks redundant here? They would never get through the has/is/get checks, or is it done simply because of performance reasons?

Personally I would like to see a guard clause here for readability:

```php
if ($reflMethod->getNumberOfRequiredParameters() > 0) {
    continue;
}
// also prevents a lot of is checks as destructor is hardly ever implemented and constructor only ever once
```

Another option would be to omit the number of required attributes at this point and only check it when needed.
```php
$attributes[lcfirst(substr($name, 3))] = 0 === $reflMethod->getNumberOfRequiredParameters();
```
This would limit the checks even more, but I'm not sure if calling the reflection method is slower or relying on `getName()` and `strpos()`

_If it comes down to preferences, feel free to ignore if you disagree_

fabpot commented:
Can you rebase? Thanks.

dunglas commented:
Rebased.

I've also refactored the code to add a guard clause according to @iltar comment but I left the constructor/destructor for now as we don't know the impact on performance such check can have.

stof commented on src/Symfony/Component/Serializer/Normalizer/ObjectNormalizer.php:
```diff
@@ -162,4 +129,62 @@ public function denormalize($data, $class, $format = null, array $context = arra
+            return self::$attributesCache[$key];
+        }
+
+        if (self::$attributesCache[$key] = $this->getAllowedAttributes($object, $context, true)) {
```
you should not assign false here. It could corrupt the cache in case an exception is thrown in the following logic, is catched, and the method is called again: as you assign false in the cache array, the cache is set and used on future calls (and breaks the contracts of the method). Use a local variable and assign the cache only on next line when using the value.

and the logic is also altered here: an empty array should be returned here. Only ``false`` triggers the fallback logic. So you need to use a strict comparison

stof replied:
btw, this means we are missing a test covering the case where the object does not have any property serialized in the groups being used (your PR switches it to serialize the whole object, which is a regression)

dunglas committed:
[Serializer] ObjectNormalizer: test normalize with a non-existing group

dunglas commented:
@stof bug fixed and test added. Thanks for the review.

fabpot commented:
Thank you @dunglas.
